src/UI/Implementation/Component/ViewControl/Renderer.php

```diff
@@ -181,7 +184,143 @@ protected function renderSortation(Component\ViewControl\Sortation $component, R
 			->withLabel($init_label);
 
 		$tpl->setVariable('SORTATION_DROPDOWN', $default_renderer->render($dd));
-	    return $tpl->get();
+		return $tpl->get();
+	}
+
+
+	protected function renderPagination(Component\ViewControl\Pagination $component, RendererInterface $default_renderer) {
+		$tpl = $this->getTemplate("tpl.pagination.html", true, true);
+
+		$component = $component->withResetSignals();
+		$triggeredSignals = $component->getTriggeredSignals();
+		if($triggeredSignals) {
+			$internal_signal = $component->getInternalSignal();
+			$signal = $triggeredSignals[0]->getSignal();
+			$component = $component->withOnLoadCode(function($id) use ($internal_signal, $signal) {
+				return "$(document).on('{$internal_signal}', function(event, signalData) {
+							il.UI.viewcontrol.pagination.onInternalSelect(event, signalData, '{$signal}', '{$id}');
+							return false;
+						})";
+			});
+
+			$id = $this->bindJavaScript($component);
+			$tpl->setVariable('ID', $id);
+		}
+
+		if(! $component->getMaxPaginationButtons()) {
+			$range = range(0, max($component->getNumberOfPages() - 1, 0));
+		} else {
+			$start = (int) ($component->getCurrentPage() - floor($component->getMaxPaginationButtons() / 2));
+			$start = max($start, 0); //0, if negative
+
+			$stop = $start + $component->getMaxPaginationButtons() - 1;
+			if($stop > $component->getNumberOfPages() - 1) {
+				$stop = max($component->getNumberOfPages() - 1, 0);
+				$start = $stop - $component->getMaxPaginationButtons();
+				$start = max($start, 0); //0, if negative
+			}
+			$range = range($start, $stop);
+
+			// If the number of pagination-buttons is limited, there should be
+			// a jump to the first and last page.
+			if(! in_array(0, $range)) {
+				$shy = $this->getPaginationShyButton(0, $component);
+				$tpl->setVariable('FIRST', $default_renderer->render($shy));
+			}
+			$last = max($component->getNumberOfPages() - 1, 0);
+			if(! in_array($last, $range)) {
+				$shy = $this->getPaginationShyButton($component->getNumberOfPages() - 1, $component);
+				$tpl->setVariable('LAST', $default_renderer->render($shy));
+			}
+		}
+
+		foreach ($range as $entry) {
+			$shy = $this->getPaginationShyButton($entry, $component);
+			if((int)$entry === $component->getCurrentPage()) {
+				$shy = $shy->withUnavailableAction();
+			}
+			$tpl->setCurrentBlock("entry");
+			$tpl->setVariable('BUTTON',	$default_renderer->render($shy));
+			$tpl->parseCurrentBlock();
+		}
+
+		$this->setPaginationRockers($component, $default_renderer, $tpl);
+		return $tpl->get();
```

Comment:
This looks rather complicated, maybe factor out some methods. This will especially help the mathy parts of that method.